fairworkflows/nanopub.py

```diff
@@ -217,16 +217,27 @@ def fetch(uri, format=Format.TRIG):
 
 
     @staticmethod
-    def rdf(assertionrdf, uri=DEFAULT_URI, introduces_concept=None, derived_from=None):
+    def rdf(assertionrdf, uri=DEFAULT_URI, introduces_concept=None, derived_from=None, attributed_to=None, nanopub_author=None):
         """
         Return the nanopub rdf, with given assertion and (defrag'd) URI, but does not sign or publish.
+        Any blank nodes in the rdf graph are replaced with the nanopub's uri, with the blank node name
+        as a fragment.
         """
 
         # Make sure passed URI is defrag'd        
+        uri = str(uri)
         uri, _ = urldefrag(uri)
-
         this_np = rdflib.Namespace(uri+'#')
 
+        # Replace blank nodes with the nanopub's uri + a fragment derived from the blank node's name
+        for s, p, o in assertionrdf:
```

Comment:
This is very hard to grasp. Maybe factor out in a separate method and add an example. I.e. what are we replacing a blancnode with exactly?

Reply:
There is a tricky thing with making these nanopubs, which is that the step or workflow (or whatever) concept that you wish to describe, should be given a URI related to that of the nanopublication. That way if someone tries to follow it, they will be taken to the nanopublication that contains its description. In other words, if the nanopub's URI is ```www.purl.org/ABC123``` then you might want the FairStep being described in its assertion graph to be ```www.purl.org/ABC123#step```.

The tricky thing, of course, is that you do not know what the nanopub URI is going to be before you build, sign and publish the nanopub. So what should you call things before that point?

My solution is to take any rdf node which should be assigned a URI based on the (published) nanopub URI, and make it a blank node (also called 'anonymous' nodes). You can give the blank node any string name you like, for example 'step' or 'workflow'. When the ```Nanopub``` module finds these blank nodes, it substitutes them with a new URI, of the form: [nanopubURI]#[blanknodename]. If blanknodename is 'step', then the URI will have a fragment '#step' after it.

If you wanted to publish several FairSteps (say) in a single nanopub, you would need to be careful that the blank nodes had different names (perhaps 'step1', 'step2', 'plan1' etc).

Does any of this make sense...?

Reply:
Yes, this makes a lot of sense! Would be good to add some if this documentation to the code. 

Maybe it would also be nice to add a test that tests this scenario where you have multiple steps (i.e. step1, step2)

Reply:
For now I've added a long comment explaining what the blank node replacement is actually doing and why.